File: project/app/src/main/java/org/owntracks/android/support/ContactImageProvider.java
package org.owntracks.android.support;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.util.Base64;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.collection.ArrayMap;
import androidx.databinding.BindingAdapter;

import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.Marker;

import org.owntracks.android.injection.qualifier.AppContext;
import javax.inject.Singleton;
import org.owntracks.android.model.FusedContact;
import org.owntracks.android.support.widgets.TextDrawable;

import java.lang.ref.WeakReference;

import javax.inject.Inject;

import timber.log.Timber;

@Singleton
public class ContactImageProvider {
    private static ContactBitmapMemoryCache memoryCache;
    private static int FACE_DIMENSIONS;


    public void invalidateCacheLevelCard(String key) {
        memoryCache.clearLevelCard(key);
    }


    private static class ContactDrawableWorkerTaskForImageView extends AsyncTask<FusedContact, Void, Bitmap> {
        final WeakReference<ImageView> target;

        ContactDrawableWorkerTaskForImageView(ImageView imageView) {
            target = new WeakReference<>(imageView);
        }

        @Override
        protected Bitmap doInBackground(FusedContact... params) {
            return getBitmapFromCache(params[0]);
        }

        protected void onPostExecute(Bitmap result) {
            if (result == null)
                return;

            ImageView imageView = target.get();
            if (imageView != null)
                imageView.setImageBitmap(result);
        }

    }

    private static class ContactDrawableWorkerTaskForMarker extends AsyncTask<FusedContact, Void, BitmapDescriptor> {
        final WeakReference<Marker> target;

        ContactDrawableWorkerTaskForMarker(Marker marker) {
            target = new WeakReference<>(marker);
        }

        @Override
        protected BitmapDescriptor doInBackground(FusedContact... params) {
            return BitmapDescriptorFactory.fromBitmap(getBitmapFromCache(params[0]));
        }

        @Override
        protected void onPostExecute(BitmapDescriptor result) {
            Marker marker = target.get();
            if (marker != null) {
                try {
                    marker.setIcon(result);
                    marker.setVisible(true);
                } catch (IllegalArgumentException e) {
                    Timber.e(e, "Error setting marker icon");
                }
            }
        }
    }

    public void setMarkerAsync(Marker marker, FusedContact contact) {
        (new ContactDrawableWorkerTaskForMarker(marker)).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, contact);
    }

    public static void setImageViewAsync(ImageView imageView, FusedContact contact) {
        //imageView.setImageDrawable(placeholder);
        (new ContactDrawableWorkerTaskForImageView(imageView)).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, contact);
    }

    @Nullable
    private static Bitmap getBitmapFromCache(@Nullable FusedContact contact) {
        Bitmap d;

        if (contact == null)
            return null;

        if (contact.hasCard()) {
            d = memoryCache.getLevelCard(contact.getId());
            if (d != null) {
                return d;
            }

            if (contact.getMessageCard().hasFace()) {
                byte[] imageAsBytes = Base64.decode(contact.getMessageCard().getFace().getBytes(), Base64.DEFAULT);
                Bitmap b = BitmapFactory.decodeByteArray(imageAsBytes, 0, imageAsBytes.length);

                if (b == null) {
                    Timber.e("Decoding card bitmap failed");
                    Bitmap fallbackBitmap = Bitmap.createBitmap(FACE_DIMENSIONS, FACE_DIMENSIONS, Bitmap.Config.ARGB_8888);
                    Canvas canvas = new Canvas(fallbackBitmap);
                    Paint paint = new Paint();
                    paint.setColor(0xFFFFFFFF);
                    canvas.drawRect(0F, 0F, (float) FACE_DIMENSIONS, (float) FACE_DIMENSIONS, paint);
                    d = getRoundedShape(fallbackBitmap);
                } else {
                    d = getRoundedShape(Bitmap.createScaledBitmap(b, FACE_DIMENSIONS, FACE_DIMENSIONS, true));
                    memoryCache.putLevelCard(contact.getId(), d);
                }
                return d;
            }
        }

        TidBitmap td = memoryCache.getLevelTid(contact.getId());
        // if cache doesn't contain a bitmap for a contact or if the cached bitmap was for an old tid, create a new one and cache it
        if (td == null || !td.isBitmapFor(contact.getTrackerId())) {
            td = new TidBitmap(contact.getTrackerId(), drawableToBitmap(TextDrawable.builder().buildRoundRect(contact.getTrackerId(), TextDrawable.ColorGenerator.MATERIAL.getColor(contact.getId()), FACE_DIMENSIONS)));
            memoryCache.putLevelTid(contact.getId(), td);
        }

        return td.getBitmap();
    }

    @Inject
    public ContactImageProvider(@AppContext Context context) {
        memoryCache = new ContactBitmapMemoryCache();
        FACE_DIMENSIONS = (int) (48 * (context.getResources().getDisplayMetrics().densityDpi / 160f));
    }

    private static class ContactBitmapMemoryCache {
        private final ArrayMap<String, Bitmap> cacheLevelCard;
        private final ArrayMap<String, TidBitmap> cacheLevelTid;

        ContactBitmapMemoryCache() {
            cacheLevelCard = new ArrayMap<>();
            cacheLevelTid = new ArrayMap<>();
        }

        synchronized void putLevelCard(String key, Bitmap value) {
            cacheLevelCard.put(key, value);
            cacheLevelTid.remove(key);
        }

        synchronized void putLevelTid(String key, TidBitmap value) {
            cacheLevelTid.put(key, value);
        }

        synchronized Bitmap getLevelCard(String key) {
            return cacheLevelCard.get(key);
        }

        synchronized TidBitmap getLevelTid(String key) {
            return cacheLevelTid.get(key);
        }

        synchronized void clear() {
            cacheLevelCard.clear();
            cacheLevelTid.clear();
        }

        synchronized void clearLevelCard(String key) {
            cacheLevelCard.remove(key);
        }

    }

    public void invalidateCache() {
        memoryCache.clear();
    }

    private static Bitmap getRoundedShape(Bitmap bitmap) {
        Bitmap output = Bitmap.createBitmap(bitmap.getWidth(), bitmap.getHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(output);

        final int color = 0xff424242;
        final Paint paint = new Paint();
        final Rect rect = new Rect(0, 0, bitmap.getWidth(), bitmap.getHeight());
        final RectF rectF = new RectF(rect);
        final float roundPx = bitmap.getWidth();

        paint.setAntiAlias(true);
        canvas.drawARGB(0, 0, 0, 0);
        paint.setColor(color);
        canvas.drawRoundRect(rectF, roundPx, roundPx, paint);

        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawBitmap(bitmap, rect, rect, paint);

        return output;
    }

    private static Bitmap drawableToBitmap(Drawable drawable) {
        if (drawable instanceof BitmapDrawable) {
            return ((BitmapDrawable) drawable).getBitmap();
        }

        int width = drawable.getIntrinsicWidth();
        width = width > 0 ? width : FACE_DIMENSIONS;
        int height = drawable.getIntrinsicHeight();
        height = height > 0 ? height : FACE_DIMENSIONS;

        Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);

        return bitmap;
    }

    @BindingAdapter({"imageProvider", "contact"})
    public static void displayFaceInViewAsync(ImageView view, Integer imageProvider, FusedContact c) {
        setImageViewAsync(view, c);
    }
}

File: project/app/src/main/java/org/owntracks/android/support/ContactImageProvider.kt
package org.owntracks.android.support

import android.content.Context
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.util.Base64
import android.widget.ImageView
import androidx.collection.ArrayMap
import androidx.databinding.BindingAdapter
import androidx.databinding.DataBindingComponent
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.Marker
import kotlinx.coroutines.*
import org.owntracks.android.injection.qualifier.AppContext
import org.owntracks.android.model.FusedContact
import org.owntracks.android.support.widgets.TextDrawable
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

internal class ContactBitmapMemoryCache {
    private val cacheLevelCard: ArrayMap<String, Bitmap?> = ArrayMap()
    private val cacheLevelTid: ArrayMap<String, TidBitmap> = ArrayMap()

    @Synchronized
    fun putLevelCard(key: String, value: Bitmap?) {
        cacheLevelCard[key] = value
        cacheLevelTid.remove(key)
    }

    @Synchronized
    fun putLevelTid(key: String, value: TidBitmap) {
        cacheLevelTid[key] = value
    }

    @Synchronized
    fun getLevelCard(key: String?): Bitmap? {
        return cacheLevelCard[key]
    }

    @Synchronized
    fun getLevelTid(key: String?): TidBitmap? {
        return cacheLevelTid[key]
    }

    @Synchronized
    fun clear() {
        cacheLevelCard.clear()
        cacheLevelTid.clear()
    }

    @Synchronized
    fun clearLevelCard(key: String?) {
        cacheLevelCard.remove(key)
    }

}

@Singleton
class ContactImageProvider @Inject constructor(@AppContext context: Context): DataBindingComponent {
    private val faceDimensions = (48 * (context.resources.displayMetrics.densityDpi / 160f)).toInt()
    fun invalidateCacheLevelCard(key: String?) {
        memoryCache.clearLevelCard(key)
    }

    fun setMarkerAsync(marker: Marker, contact: FusedContact?) {
        GlobalScope.launch(Dispatchers.Main) {
            contact?.let {
                val bitmap = BitmapDescriptorFactory.fromBitmap(getBitmapFromCache(it))
                marker.setIcon(bitmap)
                marker.isVisible = true
            }
        }
    }

    fun setImageViewAsync(imageView: ImageView, contact: FusedContact?) {
        GlobalScope.launch(Dispatchers.Main) {
            contact?.let {
                imageView.setImageBitmap(getBitmapFromCache(it))
            }
        }
    }

    private suspend fun getBitmapFromCache(contact: FusedContact?): Bitmap? {
        return withContext(Dispatchers.IO) {
            var bitmap: Bitmap?
            if (contact == null) return@withContext null
            if (contact.hasCard()) {
                bitmap = memoryCache.getLevelCard(contact.id)
                if (bitmap != null) {
                    return@withContext bitmap
                }
                if (contact.messageCard.hasFace()) {
                    val imageAsBytes = Base64.decode(contact.messageCard.face!!.toByteArray(), Base64.DEFAULT)
                    val b = BitmapFactory.decodeByteArray(imageAsBytes, 0, imageAsBytes.size)
                    if (b == null) {
                        Timber.e("Decoding card bitmap failed")
                        val fallbackBitmap = Bitmap.createBitmap(faceDimensions, faceDimensions, Bitmap.Config.ARGB_8888)
                        val canvas = Canvas(fallbackBitmap)
                        val paint = Paint()
                        paint.color = -0x1
                        canvas.drawRect(0f, 0f, faceDimensions.toFloat(), faceDimensions.toFloat(), paint)
                        bitmap = getRoundedShape(fallbackBitmap)
                    } else {
                        bitmap = getRoundedShape(Bitmap.createScaledBitmap(b, faceDimensions, faceDimensions, true))
                        memoryCache.putLevelCard(contact.id, bitmap)
                    }
                    return@withContext bitmap
                }
            }
            var td = memoryCache.getLevelTid(contact.id)
            // if cache doesn't contain a bitmap for a contact or if the cached bitmap was for an old tid, create a new one and cache it
            if (td == null || !td.isBitmapFor(contact.trackerId)) {
                td = TidBitmap(contact.trackerId, drawableToBitmap(TextDrawable.builder().buildRoundRect(contact.trackerId, TextDrawable.ColorGenerator.MATERIAL.getColor(contact.id), faceDimensions)))
                memoryCache.putLevelTid(contact.id, td)
            }
            return@withContext td.bitmap
        }

    }

    fun invalidateCache() {
        memoryCache.clear()
    }

    private fun getRoundedShape(bitmap: Bitmap): Bitmap {
        val output = Bitmap.createBitmap(bitmap.width, bitmap.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(output)
        val color = -0xbdbdbe
        val paint = Paint()
        val rect = Rect(0, 0, bitmap.width, bitmap.height)
        val rectF = RectF(rect)
        val roundPx = bitmap.width.toFloat()
        paint.isAntiAlias = true
        canvas.drawARGB(0, 0, 0, 0)
        paint.color = color
        canvas.drawRoundRect(rectF, roundPx, roundPx, paint)
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        canvas.drawBitmap(bitmap, rect, rect, paint)
        return output
    }

    private fun drawableToBitmap(drawable: Drawable): Bitmap {
        if (drawable is BitmapDrawable) {
            return drawable.bitmap
        }
        var width = drawable.intrinsicWidth
        width = if (width > 0) width else faceDimensions
        var height = drawable.intrinsicHeight
        height = if (height > 0) height else faceDimensions
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }

    @BindingAdapter("imageProvider", "contact")
    fun displayFaceInViewAsync(view: ImageView, imageProvider: Int?, c: FusedContact?) {
        setImageViewAsync(view, c);
    }

    companion object {
        private val memoryCache: ContactBitmapMemoryCache = ContactBitmapMemoryCache()
    }

    override fun getContactImageProvider(): ContactImageProvider {
        return this
    }
}